### PR TITLE
Add a way to filter out self updates from the 'updated' filters

### LINKED ISSUE #898: Add a way to filter out self updates from the 'updated' filter
e.g. some special character to indicate that self updates should be ignored

### PR DESCRIPTION
**Pull request by codemanmeet**
Fixes #898

### CODE REVIEW

File: src/main/java/backend/IssueMetadata.java [hunk 1]
```diff
@@ -12,18 +12,26 @@
...
     private final List<Comment> comments;
     private final LocalDateTime nonSelfUpdatedAt;
+    private final LocalDateTime selfUpdatedAt;
     private final int nonSelfCommentCount;
+    private final int selfCommentCount;
 
     // If isUpdated is true, nonSelfUpdatedAt will be used to sort/filter instead of updatedAt in TurboIssue
     private final boolean isUpdated;
+    private final boolean isUpdatedBySelf;
+    private final boolean isUpdatedByOthers;
```
Comment (dariusf) [inline src/main/java/backend/IssueMetadata.java]: Is there a difference between `isUpdatedByOthers` and `isUpdated`? See the comment above.

File: src/main/java/backend/IssueMetadata.java [hunk 2]
```diff
@@ -49,17 +65,27 @@ public IssueMetadata(List<TurboIssueEvent> events, List<Comment> comments) {
...
         this.events = new ArrayList<>(other.events);
         this.comments = new ArrayList<>(other.comments);
         this.nonSelfUpdatedAt = nonSelfUpdatedAt; // Calculated just prior to calling this constructor
-        this.nonSelfCommentCount = nonSelfCommentCount; // Calculated just prior to calling
+        this.selfUpdatedAt = selfUpdatedAt;
+        this.nonSelfCommentCount = nonSelfCommentCount; // Calculated just prior to calling this constructor
+        this.selfCommentCount = selfCommentCount; //Calculated just prior to calling this constructor
         this.isUpdated = true;
+        this.isUpdatedByOthers = isUpdatedByOthers; //Calculated prior to calling this constructor
+        this.isUpdatedBySelf = isUpdatedBySelf; // Calculated prior to calling this constructor
```
Comment (dariusf) [inline src/main/java/backend/IssueMetadata.java]: I don't think these comments are necessary. The constructor receives these values as parameters so it's kind of a given.

File: src/main/java/backend/Logic.java [hunk 3]
```diff
@@ -157,32 +160,57 @@ public void refresh() {
...
+                    isUpdatedBySelf = true;
                 }
             }
 
             entry.setValue(new IssueMetadata(currentMetadata,
                     LocalDateTime.ofInstant(lastNonSelfUpdate.toInstant(), ZoneId.systemDefault()),
-                    calculateNonSelfCommentCount(currentMetadata.getComments(), currentUser)
+                    LocalDateTime.ofInstant(lastSelfUpdate.toInstant(), ZoneId.systemDefault()),
+                    calculateCommentCount(currentMetadata.getComments(), currentUser, IssueType.OTHER_UPDATED),
+                    calculateCommentCount(currentMetadata.getComments(), currentUser, IssueType.SELF_UPDATED),
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: All this logic should be in the `IssueMetadata` constructor (or some factory method), and broken down further into different methods. This method is doing too much, plus `IssueMetadata` should handle its own construction.

File: src/main/java/backend/Logic.java [hunk 4]
```diff
@@ -157,32 +160,57 @@ public void refresh() {
...
+            for (Comment comment : comments) {
+                if (!comment.getUser().getLogin().equalsIgnoreCase(currentUser)) {
+                    result++;
+                }
+            }
+        } else if (issueType == IssueType.SELF_UPDATED){
+            for (Comment comment : comments) {
+                if (comment.getUser().getLogin().equalsIgnoreCase(currentUser)) {
+                    result++;
+                }
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: The conditional logic should be factored out into a predicate so we don't have to repeat ourselves. This may also be formulated more clearly in terms of a `filter`, then `count`, but we can try the predicate first.

File: src/main/java/filter/expression/Qualifier.java [hunk 5]
```diff
@@ -25,9 +25,14 @@
 public class Qualifier implements FilterExpression {
 
     public static final String UPDATED = "updated";
+    public static final String UPDATED_OTHERS = "updatedByOthers";
     public static final String REPO = "repo";
     public static final String SORT = "sort";
 
+    private enum IssueType{
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: `IssueType` is too generic a name. How about `UpdatedKind`, to make it more consistent with how it's named below?

File: src/main/java/filter/expression/Qualifier.java [hunk 6]
```diff
@@ -35,7 +40,8 @@
         "date", "nonSelfUpdate", "desc", "description", "has", "id", "in", "involves",
         "is", "issue", "keyword", "label", "labels", "merged", "milestone", "milestones",
         "no", "open", "pr", "pullrequest", "read", "repo", "sort", "state", "status",
-        "title", "type", "unmerged", "unread", "updated", "user"
+        "title", "type", "unmerged", "unread", "updated", "updatedByOthers",
+        "updatedBySelf", "user"
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: It'd be good if we could put `nonSelfUpdate` in alphabetical position as well.

File: src/main/java/filter/expression/Qualifier.java [hunk 7]
```diff
@@ -527,17 +537,37 @@ private boolean satisfiesUpdatedHours(TurboIssue issue) {
...
+                    hoursSinceUpdate = Utility.safeLongToInt(issue.getMetadata().getNonSelfUpdatedAt()
+                            .until(getCurrentTime(), ChronoUnit.HOURS));
+                }
+                break;
+            case ALL_UPDATED:
+                if (issue.getMetadata().isUpdated()) {
+                    // Second time being filtered, we now have metadata from source, so we can use getNonSelfUpdatedAt.
+                    hoursSinceUpdate = Utility.safeLongToInt(issue.getUpdatedAt()
+                            .until(getCurrentTime(), ChronoUnit.HOURS));
+                }
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: We should use the conditional to decide the date field to use, instead of the hours. The hours can be computed based on the date field after that.

Also, the comments are duplicated.

File: src/main/java/filter/expression/Qualifier.java [hunk 8]
```diff
@@ -527,17 +537,37 @@ private boolean satisfiesUpdatedHours(TurboIssue issue) {
...
         }
 
+        // First time being filtered (haven't gotten metadata from source yet).
+        if (hoursSinceUpdate == Integer.MIN_VALUE) {
+            hoursSinceUpdate = Utility.safeLongToInt(issue.getUpdatedAt().until(getCurrentTime(), ChronoUnit.HOURS));
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: This case can be placed first, since it's shorter. Same issue -- we should decide the date instead of the hours.

File: src/main/java/ui/listpanel/ListPanel.java [hunk 9]
```diff
@@ -60,7 +60,9 @@ public ListPanel(UI ui, IModel model, PanelControl parentPanelControl, int panel
      * @return true if the issue has changed, false otherwise
      */
     private boolean issueHasNewComments(TurboIssue issue, boolean hasMetadata) {
-        if (currentFilterExpression.getQualifierNames().contains(Qualifier.UPDATED) && hasMetadata) {
+        List<String> filterQualifierNames = currentFilterExpression.getQualifierNames();
+        if ((filterQualifierNames.contains(Qualifier.UPDATED) ||
```
Comment (dariusf) [inline src/main/java/ui/listpanel/ListPanel.java]: Do we need to check for the self update qualifier as well?

`hasMetadata` could be put first the boolean expression. The presence of qualifiers can be checked into a static `Qualifier` helper method.

File: src/test/java/tests/FilterEvalTests.java [hunk 10]
```diff
@@ -425,19 +425,39 @@ public void updated() {
         issue.setUpdatedAt(now.minusDays(2));
 
         assertEquals(false, matches("updated:<24", issue));
+        assertEquals(false, matches("updatedByOthers:<24", issue));
+        assertEquals(false, matches("updatedBySelf:<24", issue));
```
Comment (dariusf) [inline src/test/java/tests/FilterEvalTests.java]: These tests don't cover what was changed. You have to stub out each the test issue's metadata with events and comments, and ensure that the right qualifier matches in the right case. There are 2^3 cases to cover.

File: src/main/java/ui/listpanel/ListPanel.java (hunk 9 above)
Comment (codemanmeet) [inline src/main/java/ui/listpanel/ListPanel.java]: Nope, we don't. We care about comments/events from other users in this context.

File: src/main/java/filter/expression/Qualifier.java (hunk 5 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Changed.

File: src/test/java/tests/FilterEvalTests.java (hunk 10 above)
Comment (codemanmeet) [inline src/test/java/tests/FilterEvalTests.java]: I added tests to check for `updatedByOthers` and `updatedBySelf` separately with the `IssueMetadata` construction to indicate when the issue was last self updated or last updated by others.

File: src/main/java/filter/expression/Qualifier.java (hunk 6 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Changed.

File: src/main/java/ui/listpanel/ListPanel.java (hunk 9 above)
Comment (dariusf) [inline src/main/java/ui/listpanel/ListPanel.java]: Okay. I guess you could add a comment explaining this, since you're purposefully omitting the self check.

File: src/main/java/backend/IssueMetadata.java (hunk 2 above)
Comment (codemanmeet) [inline src/main/java/backend/IssueMetadata.java]: Addressed!

File: src/main/java/backend/IssueMetadata.java (hunk 1 above)
Comment (codemanmeet) [inline src/main/java/backend/IssueMetadata.java]: Yup, I included a difference. The previous implementation of the updated filter was causing some updates to go missed. There's the issue of cache in doing the filtering with `nonSelfUpdatedAt` after metadata is available versus always doing with `updated` whether or not `metadata` is available. The issue has been addressed in #912

File: src/main/java/backend/Logic.java (hunk 3 above)
Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: Addressed!

File: src/main/java/ui/listpanel/ListPanel.java (hunk 9 above)
Comment (codemanmeet) [inline src/main/java/ui/listpanel/ListPanel.java]: Sure.

File: src/main/java/filter/expression/Qualifier.java (hunk 7 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Addressed!

File: src/main/java/filter/expression/Qualifier.java (hunk 8 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Addressed!

**Comment by damithc** [comment]
Should there be some changes to documentation?

**Comment by codemanmeet** [comment]
Yup, will add those as well.

File: src/main/java/backend/IssueMetadata.java (hunk 1 above)
Comment (dariusf) [inline src/main/java/backend/IssueMetadata.java]: So `isUpdated` tells us whether or not metadata is up to date in general, and the other two tell us whom the last update was by?

Maybe you could add a brief comment explaining the purpose of each field. Without context it can be difficult to figure out what they are for.

File: src/main/java/backend/Logic.java [hunk 11]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
...
+                    isUpdatedBySelf = true;
                 }
             }
 
             entry.setValue(new IssueMetadata(currentMetadata,
                     LocalDateTime.ofInstant(lastNonSelfUpdate.toInstant(), ZoneId.systemDefault()),
-                    calculateNonSelfCommentCount(currentMetadata.getComments(), currentUser)
+                    LocalDateTime.ofInstant(lastSelfUpdate.toInstant(), ZoneId.systemDefault()),
+                    calculateCommentCount(currentMetadata.getComments(), currentUser, UpdatedKind.OTHER_UPDATED),
+                    calculateCommentCount(currentMetadata.getComments(), currentUser, UpdatedKind.SELF_UPDATED),
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: Same issue as before. The values of these two arguments can be derived from the others, so you can compute them in the `TurboIssue` constructor. We want to keep `Logic` as high-level as possible.

File: src/main/java/backend/Logic.java [hunk 12]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
...
             Date lastNonSelfUpdate = new Date(0);
+            Date lastSelfUpdate = new Date(0);
+            boolean isUpdatedByOthers = false;
+            boolean isUpdatedBySelf = false;
             for (TurboIssueEvent event : currentMetadata.getEvents()) {
-                if (!event.getActor().getLogin().equalsIgnoreCase(currentUser)
-                        && event.getDate().after(lastNonSelfUpdate)) {
+                if (isNewEventByOther(currentUser, event, lastNonSelfUpdate)) {
                     lastNonSelfUpdate = event.getDate();
+                    isUpdatedByOthers = true;
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: All this logic is about the (re-)construction of `IssueMetadata` given the current user (which isn't available deeper in the back end). Could it be moved into a new `IssueMetadata` constructor? `new IssueMetadata(existingMetadata, currentUser)`

File: src/main/java/backend/Logic.java [hunk 13]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
         for (Map.Entry<Integer, IssueMetadata> entry : metadata.entrySet()) {
             IssueMetadata currentMetadata = entry.getValue();
             Date lastNonSelfUpdate = new Date(0);
+            Date lastSelfUpdate = new Date(0);
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: You should use `LocalDateTime.now()` for these. `Date` is a legacy class and could only be used for compatibility.

File: src/main/java/backend/Logic.java [hunk 14]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
...
+                if (isNewCommentBySelf(currentUser, comment, lastSelfUpdate)){
+                    lastSelfUpdate = comment.getCreatedAt();
+                    isUpdatedBySelf = true;
                 }
             }
 
             entry.setValue(new IssueMetadata(currentMetadata,
                     LocalDateTime.ofInstant(lastNonSelfUpdate.toInstant(), ZoneId.systemDefault()),
-                    calculateNonSelfCommentCount(currentMetadata.getComments(), currentUser)
+                    LocalDateTime.ofInstant(lastSelfUpdate.toInstant(), ZoneId.systemDefault()),
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: `LocalDateTime` objects are immutable, so you don't need to clone them.

File: src/main/java/backend/Logic.java [hunk 15]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
...
+                    calculateCommentCount(currentMetadata.getComments(), currentUser, UpdatedKind.SELF_UPDATED),
+                    isUpdatedByOthers,
+                    isUpdatedBySelf
             ));
         }
         return metadata;
     }
 
-    private int calculateNonSelfCommentCount(List<Comment> comments, String currentUser) {
+    private boolean isNewCommentByOther(String currentUser, Comment comment, Date lastNonSelfUpdate){
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: All these helper methods can be static.

File: src/main/java/backend/Logic.java [hunk 16]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
...
         int result = 0;
-        for (Comment comment : comments) {
-            if (!comment.getUser().getLogin().equalsIgnoreCase(currentUser)) {
-                result++;
+        if (updatedKind == UpdatedKind.OTHER_UPDATED) {
+            for (Comment comment : comments) {
+                if (!isCommentBySelf(currentUser, comment)) {
+                    result++;
+                }
+            }
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: `result` can be eliminated by filtering instead.
```java
return comments.stream()
    .filter(c -> !isCommentBySelf(currentUser, c))
    .count();
```

File: src/main/java/backend/Logic.java [hunk 17]
```diff
@@ -157,37 +160,84 @@ public void refresh() {
...
+            for (Comment comment : comments) {
+                if (isCommentBySelf(currentUser, comment)) {
+                    result++;
+                }
             }
         }
         return result;
     }
 
+    private boolean isCommentBySelf(String currentUser, Comment comment){
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: Can also be static.

File: src/main/java/filter/expression/Qualifier.java [hunk 18]
```diff
@@ -527,17 +538,33 @@ private boolean satisfiesUpdatedHours(TurboIssue issue) {
             return false;
         }
 
-        int hoursSinceUpdate;
+        int hoursSinceUpdate = Integer.MIN_VALUE;
+        LocalDateTime dateOfUpdate = LocalDateTime.ofEpochSecond(0, 0, ZoneOffset.ofHours(0));
+        // First time being filtered (haven't gotten metadata from source yet).
+        if (hoursSinceUpdate == Integer.MIN_VALUE) {
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: Moving this up might have caused a bug -- this is always true. Also you might have intended to return early.

File: src/main/java/ui/listpanel/ListPanel.java [hunk 19]
```diff
@@ -71,6 +72,12 @@ private boolean issueHasNewComments(TurboIssue issue, boolean hasMetadata) {
         }
     }
 
+    private static boolean containsUpdatedQualifier(List<String> filterQualifierNames){
```
Comment (dariusf) [inline src/main/java/ui/listpanel/ListPanel.java]: This method and the fields above could go in `Qualifier`. The name is a bit too generic for what it does also.

File: src/test/java/tests/FilterEvalTests.java [hunk 20]
```diff
@@ -433,6 +434,38 @@ public void created() {
     }
 
     @Test
+    public void updatedByOther(){
+        LocalDateTime now = LocalDateTime.now();
```
Comment (dariusf) [inline src/test/java/tests/FilterEvalTests.java]: We can probably merge these two tests, since they're testing very similar things. You don't have to test ranges, since they are covered elsewhere. Something more important is which qualifiers match in which scenarios: given an issue, there are 4 possibilities for whether the two qualifiers match, so that would be more useful to focus on.

```java
TurboIssue issue = ...;
assertEquals(false, matches("updatedByOthers:24", issue));
assertEquals(false, matches("updatedBySelf:24", issue));
// repeat for (true, true), (false, true), (true, false)
```
There's also the cases where issues can be updated by **both** others and self, or **none**. A test to demonstrate which takes precedence would be useful.

File: src/main/java/backend/IssueMetadata.java (hunk 1 above)
Comment (codemanmeet) [inline src/main/java/backend/IssueMetadata.java]: Yup, exactly.

File: src/main/java/backend/Logic.java (hunk 13 above)
Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: I have to use `Date` because `event.getDate()` and `comment.getCreatedAt()` return `Date` instead of `LocalDateTime` objects. This would require a lot of refactoring of old code again. We can add that refactoring part in a separate issue.

File: src/main/java/backend/Logic.java (hunk 14 above)
Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: Same as above. The above code basically converts the `Date` object to `LocalDateTime` object.

File: src/main/java/backend/Logic.java (hunk 17 above)
Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: Changed.

File: src/main/java/backend/Logic.java (hunk 12 above)
Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: Done!

File: src/main/java/backend/Logic.java (hunk 16 above)
Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: Done!

Comment (codemanmeet) [inline src/main/java/backend/Logic.java]: I found this really useful!

File: src/main/java/filter/expression/Qualifier.java (hunk 18 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Fixed.

File: src/main/java/ui/listpanel/ListPanel.java (hunk 19 above)
Comment (codemanmeet) [inline src/main/java/ui/listpanel/ListPanel.java]: Fixed.

File: src/test/java/tests/FilterEvalTests.java (hunk 20 above)
Comment (codemanmeet) [inline src/test/java/tests/FilterEvalTests.java]: Fixed. Added all the possible test cases.

File: docs/filters.md [hunk 21]
```diff
@@ -108,6 +108,8 @@ This will not work for ambiguous expressions (containing OR or NOT operators) an
 - [`is`](#is)
 - [`created`](#created)
 - [`updated`](#updated)
+- [`updatedByOthers`](#updatedByOthers)
+- [`updatedBySelf`](#updatedBySelf)
```
Comment (damithc) [inline docs/filters.md]: Didn't we decide to use `-word` format instead, which is similar to git command format?

File: docs/filters.md [hunk 22]
```diff
@@ -222,6 +224,18 @@ Matches issues which were created on a given date, or within a given date range.
 
 Matches issues which were updated in the given number of hours. For example, `updated:<24` would match issues updated in the last day. If a number `n` is given, it is implicitly translated to `<n`.
 
+### updatedByOthers
+
+*Expects a number or  number range*
+
+Matches issues which were updated in the given number of hours by users other than the logged in user. For example, `updated:<24` would match issues updated in the last day by other users. If a number `n` is given, it is implicitly translated to `<n`.
+
```
Comment (damithc) [inline docs/filters.md]: What's meant by 'number range'? Something like `2-4`? In that case, give an example.

File: src/main/java/backend/IssueMetadata.java [hunk 23]
```diff
@@ -4,26 +4,39 @@
...
+    private final int selfCommentCount;
 
-    // If isUpdated is true, nonSelfUpdatedAt will be used to sort/filter instead of updatedAt in TurboIssue
-    private final boolean isUpdated;
+    private final boolean isUpdated; //for any update
+    private final boolean isUpdatedBySelf; // for update by self
+    private final boolean isUpdatedByOthers; // for update by others
+
+    private enum UpdatedKind {
+        SELF_UPDATED, OTHER_UPDATED
```
Comment (damithc) [inline src/main/java/backend/IssueMetadata.java]: `UPDATED_BY_SELF, UPDATED_BY_OTHERS` ?

File: docs/filters.md (hunk 22 above)
Comment (damithc) [inline docs/filters.md]: Or we can refer to some other place where more examples of date ranges are given. No need to repeat everywhere.

File: docs/filters.md (hunk 21 above)
Comment (dariusf) [inline docs/filters.md]: Also these links don't work. Seems to be case-sensitive.

File: docs/filters.md (hunk 22 above)
Comment (dariusf) [inline docs/filters.md]: The format for number ranges is given at the top. Maybe in future we can add links pointing back up.

I agree about the duplication. Since these come right after `updated`, it should be enough to just point out the differences in behaviour.

File: src/main/java/backend/IssueMetadata.java [hunk 24]
```diff
@@ -49,17 +70,87 @@ public IssueMetadata(List<TurboIssueEvent> events, List<Comment> comments) {
...
+            if (isNewCommentBySelf(currentUser, comment, lastSelfUpdate)){
+                lastSelfUpdate = comment.getCreatedAt();
+                isUpdatedBySelf = true;
+            }
+        }
+
+        this.events = new ArrayList<>(existingMetadata.events);
+        this.comments = new ArrayList<>(existingMetadata.comments);
+        this.nonSelfUpdatedAt = LocalDateTime.ofInstant(lastNonSelfUpdate.toInstant(), ZoneId.systemDefault());
+        this.selfUpdatedAt = LocalDateTime.ofInstant(lastSelfUpdate.toInstant(), ZoneId.systemDefault());
```
Comment (dariusf) [inline src/main/java/backend/IssueMetadata.java]: Use `Utility.dateToLocalDateTime` instead.

File: src/main/java/backend/IssueMetadata.java [hunk 25]
```diff
@@ -49,17 +70,87 @@ public IssueMetadata(List<TurboIssueEvent> events, List<Comment> comments) {
...
+        return !event.getActor().getLogin().equalsIgnoreCase(currentUser)
+                && event.getDate().after(lastNonSelfUpdate);
+    }
+
+    private static int calculateCommentCount(List<Comment> comments, String currentUser, UpdatedKind updatedKind) {
+        int result = 0;
+        if (updatedKind == UpdatedKind.OTHER_UPDATED) {
+            return (int) (comments.stream().filter(c -> !isCommentBySelf(currentUser, c)).count());
+        } else if (updatedKind == UpdatedKind.SELF_UPDATED){
+            return (int) (comments.stream().filter(c -> isCommentBySelf(currentUser, c)).count());
```
Comment (dariusf) [inline src/main/java/backend/IssueMetadata.java]: Use `Utility.safeLongToInt` instead.

File: src/main/java/filter/expression/Qualifier.java [hunk 26]
```diff
@@ -527,18 +538,34 @@ private boolean satisfiesUpdatedHours(TurboIssue issue) {
...
+                }
+                break;
+            case OTHER_UPDATED:
+                if (issue.getMetadata().isUpdatedByOthers()) {
+                    dateOfUpdate = issue.getMetadata().getNonSelfUpdatedAt();
+                }
+                break;
+            case ALL_UPDATED:
+                // The or condition for dateOfUpdate is for first time check
+                if (issue.getMetadata().isUpdated() || dateOfUpdate == null) {
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: This condition is again always true. Potential bug.

File: src/main/java/filter/expression/Qualifier.java [hunk 27]
```diff
@@ -527,18 +538,34 @@ private boolean satisfiesUpdatedHours(TurboIssue issue) {
...
+                }
         }
 
-        return updatedRange.encloses(hoursSinceUpdate);
+        if (dateOfUpdate != null){
+            int hoursSinceUpdate = Utility.safeLongToInt(dateOfUpdate.until(getCurrentTime(), ChronoUnit.HOURS));
+            return updatedRange.encloses(hoursSinceUpdate);
+        } else {
+            return false;
+        }
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: It'd be good to switch the order of these branches, so when reading the easier condition comes first.

File: src/main/java/ui/listpanel/ListPanel.java [hunk 28]
```diff
@@ -60,8 +60,16 @@ public ListPanel(UI ui, IModel model, PanelControl parentPanelControl, int panel
      * @param issue
      * @return true if the issue has changed, false otherwise
      */
+
+    private static boolean qualifierNamesHaveUpdatedQualifier(List<String> filterQualifierNames){
+        // omitting self-updated issues since we only care about updated comments from other users.
+        return (filterQualifierNames.contains(Qualifier.UPDATED) ||
+                filterQualifierNames.contains(Qualifier.UPDATED_OTHERS));
+    }
```
Comment (dariusf) [inline src/main/java/ui/listpanel/ListPanel.java]: This should be moved to `Qualifier`, along with the code that is calling it. The whole thing can take in a `FilterExpression` and return a boolean value.

File: src/main/java/filter/expression/Qualifier.java (hunk 27 above)
Comment (damithc) [inline src/main/java/filter/expression/Qualifier.java]: Yup, as a rule of thumb, 'pick the lower hanging fruits first' to improve readability.

File: src/test/java/tests/FilterEvalTests.java [hunk 29]
```diff
@@ -433,6 +434,59 @@ public void created() {
...
 
     @Test
+    public void updatedBySelfAndOther(){
+        LocalDateTime now = LocalDateTime.now();
+        Qualifier.setCurrentTime(now);
+        TurboIssue issueUpdated = new TurboIssue(REPO, 1, "");
+        IssueMetadata updatedByOtherMetadata = new IssueMetadata(issueUpdated.getMetadata(), now.minusHours(4),
+                now.minusHours(4), 2, 2, true, false); //UpdatedByOther true with updated time 4 hours ago.
+        issueUpdated.setMetadata(updatedByOtherMetadata);
+        assertEquals(false, matches("updatedByOthers:<2", issueUpdated));
```
Comment (damithc) [inline src/test/java/tests/FilterEvalTests.java]: Why 2? is that a boundary value?

File: src/main/java/filter/expression/Qualifier.java (hunk 26 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Yup figured the `if` condition is not even required. Just obtain the `updatedDate` and check whether it encloses the range.

File: src/test/java/tests/FilterEvalTests.java (hunk 29 above)
Comment (codemanmeet) [inline src/test/java/tests/FilterEvalTests.java]: Wanted to check for a value less than 4. I also changed tests to include 4 as the boundary.

File: src/main/java/filter/expression/Qualifier.java [hunk 30]
```diff
@@ -20,17 +21,23 @@
 public class Qualifier implements FilterExpression {
 
     public static final String UPDATED = "updated";
+    public static final String UPDATED_OTHERS = "updatedByOthers";
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: This isn't updated to reflect the new convention.

File: docs/filters.md [hunk 31]
```diff
@@ -220,7 +222,19 @@ Matches issues which were created on a given date, or within a given date range.
...
+*Expects a number or  number range*
+
+Matches issues which were updated in the given number of hours by users other than the logged in user. For example, `updated:<24` would match issues updated in the last day by other users. Number ranges are written using a relational operator (.e.g `>5`, `<=10`). If a number `n` is given, it is implicitly translated to `<n`.
+
+### updated-self
+
+*Expects a number or  number range*
+
+Matches issues which were updated in the given number of hours by the logged in user. For example, `updated:<24` would match issues updated in the last day by the logged in user. Number ranges are written using a relational operator (.e.g `>5`, `<=10`). If a number `n` is given, it is implicitly translated to `<n`.
```
Comment (dariusf) [inline docs/filters.md]: The descriptions for `updated-others` and `updated-self` still refer to `updated`.

Also, no need to duplicate the examples, just say something like, 'Like `updated`, but only matches issues updated by other users/the current user'.

File: src/main/java/backend/IssueMetadata.java [hunk 32]
```diff
@@ -3,27 +3,40 @@
...
     private final LocalDateTime nonSelfUpdatedAt;
+    private final LocalDateTime selfUpdatedAt;
     private final int nonSelfCommentCount;
+    private final int selfCommentCount;
 
-    // If isUpdated is true, nonSelfUpdatedAt will be used to sort/filter instead of updatedAt in TurboIssue
-    private final boolean isUpdated;
+    private final boolean isUpdated; //for any update
+    private final boolean isUpdatedBySelf; // for update by self
+    private final boolean isUpdatedByOthers; // for update by others
```
Comment (dariusf) [inline src/main/java/backend/IssueMetadata.java]: Sorry, this again... is `isUpdated` equal to `isUpdatedBySelf || isUpdatedByOthers`? Like, is there a case where the values of these can differ? If not, we can get rid of this field and implement `isUpdated()` in terms of the other ones.

File: src/main/java/ui/listpanel/ListPanel.java (hunk 28 above)
Comment (dariusf) [inline src/main/java/ui/listpanel/ListPanel.java]: This bit is still relevant I think. We want to minimise operations on `FilterExpression`s outside `Qualifier` and the subclasses themselves.

File: src/test/java/tests/FilterLexerTests.java [hunk 33]
```diff
@@ -37,8 +37,8 @@ public void lexer() {
             new Lexer("a,b").lex();
         } catch (ParseException ignored) {}
 
-        assertEquals(new Lexer("sort: a, b ").lex(), Arrays.asList(
-            new Token(TokenType.QUALIFIER, "sort:"),
+        assertEquals(new Lexer("sort-self: a, b ").lex(), Arrays.asList(
+            new Token(TokenType.QUALIFIER, "sort-self:"),
```
Comment (dariusf) [inline src/test/java/tests/FilterLexerTests.java]: Don't change this test, just add another. You can also cover more pathological cases, like `-aaa-bbb`, `---aaa-bbb`, `aaa-bbb-`, etc.

**Comment by dariusf** [comment]
Last round. I have a feeling that uncovered line is a false positive, so will merge once this batch is done and CI passes.

**Comment by damithc** [comment]
Wasn't there some change to sort command? Was the documentation updated for it?

**Comment by dariusf** [comment]
The tests just happen to contain the word 'sort'. No changes to functionality.

**Comment by damithc** [comment]
Oh? We didn't change `sort:nonSelfUpdate`?

**Comment by dariusf** [comment]
No, it still uses camel case.

**Comment by damithc** [comment]
I see. Let's change it to be consistent. Can be done separately.

**Comment by acruis** [comment]
@codemanmeet Is `updated-others` supposed to be used in conjunction with the original `updated` filter?

**Comment by codemanmeet** [comment]
Nope it doesn't have to be used in conjunction.
